### PR TITLE
Evaluation worker

### PR DESCRIPTION
**Pull request by 0xDiBa**
Mozilla compliance evaluation worker is, hopefully :smile:, ready.

### CODE REVIEW

**Comment by jvehent** [comment]
I added a few unit tests to the worker to verify that we get the right levels, and it seems that the tests are failing. I have not yet had time to look through the worker code, but you can reproduce using
```
$ go test github.com/mozilla/tls-observatory/worker/mozillaEvaluationWorker/
--- FAIL: TestLevels (0.00s)
	mozillaEvaluationWorker_test.go:56: Measured level bad does not match expected of modern
	mozillaEvaluationWorker_test.go:56: Measured level bad does not match expected of intermediate
	mozillaEvaluationWorker_test.go:56: Measured level bad does not match expected of old
FAIL
FAIL	github.com/mozilla/tls-observatory/worker/mozillaEvaluationWorker	0.003s
```

**Comment by 0xDiBa** [comment]
I will take a look at it later tonight.
It shouldn't be something big.

**Comment by 0xDiBa** [comment]
Fixed some bugs in the evaluator functions.
I was breaking compliance when configurations were not implementing ocsp and/or serverside ordering was not enforced.

Current test output is :
```
$ go test
--- FAIL: TestLevels (0.00s)
	mozillaEvaluationWorker_test.go:56: Measured level bad does not match expected of modern
	mozillaEvaluationWorker_test.go:56: Measured level intermediate with bad ordering does not match expected of intermediate
	mozillaEvaluationWorker_test.go:56: Measured level old with bad ordering does not match expected of old
FAIL
exit status 1
FAIL	github.com/mozilla/tls-observatory/worker/mozillaEvaluationWorker	0.005s
```

I am testing compliance against the configurations that can be found [here](https://github.com/mozilla/server-side-tls/blob/4.0/ServerSideTLSConfigurations.json).

I debugged each output and the results are fine.
( "modern" input gives bad result because in the json only ecdsa-with-SHA256 -and not sha256WithRSAEncryption- is considered a modern signature, the other two are self-explanatory)

The tests are, reasonably enough, failing. 
We should adapt either the test ( to reflect configurations that are created according to the json file ) or the json ( to reflect the configurations we want to achieve ) to the other one.

File: worker/mozillaEvaluationWorker/mozillaEvaluationWorker.go [hunk 1]
```diff
@@ -0,0 +1,550 @@
...
+			}
+		}
+
+		if cs.PFS != "None" {
+			if !hasGoodPFS(cs.PFS, 2048, 256, false) {
+				hasPFS = false
+			}
+		}
+
+		if cs.SigAlg != modern.CertificateSignature {
```
Comment (jvehent) [inline worker/mozillaEvaluationWorker/mozillaEvaluationWorker.go]: `sigalg` represents the signature algorithm used by a TLS connection. It is not related to the signature of the certificate used by the connection, so this comparison will always fail.
That means the evaluation worker need to evaluate certificates in addition to connections.

Needs to be changed for all 3 levels.

Comment (0xDiBa) [inline worker/mozillaEvaluationWorker/mozillaEvaluationWorker.go]: Hadn't realised that.
Going to change it right away and hopefully we will finish with the worker.

**Comment by jvehent** [comment]
This looks fine and could be deployed modulo fixing the issue with `sigalg`. I can help but wonder if it could be refactored further to not have any hardcoded tests (eg. `if !hasSSLv3` and so on), and just do tests based on the content of the json file. Maybe for a v2?

**Comment by 0xDiBa** [comment]
If the structure and contents of the json file is finalised we can spend a little more time to make the worker "configuration agnostic", which will allow us to change compliance checking parameters on the fly without redeploying the code ( by getting the configuration parameters directly from the json hosted online ).

In case the json and configuration specific parameters need to be worked out, I would suggest we deploy to have a fully working ( as promised ) framework and make v2 of the evaluation worker on a next iteration.

**Comment by jvehent** [comment]
And it passes the tests and builds! I'm deploying it in stage now for testing, and will do prod tomorrow. Thanks!

**Comment by jvehent** [comment]
Deployed the new scanner, but the analysis table does not get filled up. Something missing in the code or bug on my side?
```
observatory=> select * from analysis;
 id | scan_id | worker_name | output 
----+---------+-------------+--------
(0 rows)
```

**Comment by 0xDiBa** [comment]
That's what you get for not adding all files to commit. :wink:
Chose to have the worker imports in a separate file in the main package of tlsobs-scanner.

**Comment by 0xDiBa** [comment]
Oh and as you understood it was clearly a bug on your side..

**Comment by jvehent** [comment]
I'm still not getting results in the analysis table. I'm fixing a few things around the scanner, and improving logging with the hope to find the issue. Have you successfully ran this evaluation code in your own scanner?

Side note: from my reading of the code, the results api endpoint does not return analysis data. It think it should be added there, in the form of an array of analysis. Thoughts?

**Comment by jvehent** [comment]
Ha, now I get worker analysis in the database, but a bug remain in the percentage updating code (it never reaches 100%). Progress, getting there.
What about the api endpoint to return analysis results?

**Comment by 0xDiBa** [comment]
I was getting results in the analysis table with what we had yesterday.
That was strange.

You are right about the results API,
I am going to add an analysis field in the scan struct and populate it with the available worker results.
Shouldn't take long I hope.

**Comment by jvehent** [comment]
Redeployed to stage and working properly. Let's leave this open to add the API endpoint and update the `tlsobs` client to display analysis reports (I'll do that part).

**Comment by jvehent** [comment]
Do you intentionally return the analysis results base64 encoded?
```
    "analysis": [
        {
            "analyzer": "mozillaEvaluationWorker",
            "id": 4,
            "result": "eyJsZXZlbCI6ICJpbnRlcm1lZGlhdGUgd2l0aCBiYWQgb3JkZXJpbmciLCAiZmFpbHVyZXMiOiB7ImJhZCI6IG51bGwsICJvbGQiOiBbInNoYTI1NldpdGhSU0FFbmNyeXB0aW9uIGlzIG5vdCBhbiBvbGQgc2lnbmF0dXJlIiwgImNvbnNpZGVyIGVuYWJsaW5nIFNTTHYzIiwgImFkZCBTU0x2MyBzdXBwb3J0IiwgImNvbnNpZGVyIGVuYWJsaW5nIE9DU1Agc3RhcGxpbmciLCAiaXQgaXMgcmVjb21tZW5kZWQgdG8gdXNlIHNoYTEiXSwgIm1vZGVybiI6IFsic2hhMjU2V2l0aFJTQUVuY3J5cHRpb24gaXMgbm90IGEgbW9kZXJuIHNpZ25hdHVyZSIsICJyZW1vdmUgRUNESEUtUlNBLUFFUzEyOC1TSEEgY2lwaGVyIiwgInJlbW92ZSBESEUtUlNBLUFFUzEyOC1TSEEgY2lwaGVyIiwgInJlbW92ZSBFQ0RIRS1SU0EtQUVTMjU2LVNIQSBjaXBoZXIiLCAicmVtb3ZlIEFFUzEyOC1HQ00tU0hBMjU2IGNpcGhlciIsICJyZW1vdmUgQUVTMTI4LVNIQTI1NiBjaXBoZXIiLCAicmVtb3ZlIEFFUzEyOC1TSEEgY2lwaGVyIiwgInJlbW92ZSBBRVMyNTYtR0NNLVNIQTM4NCBjaXBoZXIiLCAicmVtb3ZlIEFFUzI1Ni1TSEEyNTYgY2lwaGVyIiwgInJlbW92ZSBBRVMyNTYtU0hBIGNpcGhlciIsICJyZW1vdmUgREVTLUNCQzMtU0hBIGNpcGhlciIsICJkaXNhYmxlIFRMU3YxIHByb3RvY29sIiwgImNvbnNpZGVyIGVuYWJsaW5nIE9DU1Agc3RhcGxpbmciLCAiaXQgaXMgcmVjb21tZW5kZWQgdG8gdXNlIHNoYTI1NiIsICJ1c2UgREhFIG9mIGF0IGxlYXN0IDIwNDhiaXRzIGFuZCBFQ0Mgb2YgYXQgbGVhc3QgMjU2Yml0cyJdLCAiaW50ZXJtZWRpYXRlIjogWyJzaGEyNTZXaXRoUlNBRW5jcnlwdGlvbiBpcyBub3QgYW4gaW50ZXJtZWRpYXRlIHNpZ25hdHVyZSIsICJjb25zaWRlciBlbmFibGluZyBPQ1NQIHN0YXBsaW5nIiwgIml0IGlzIHJlY29tbWVuZGVkIHRvIHVzZSBzaGEyNTYiLCAidXNlIERIRSBvZiBhdCBsZWFzdCAyMDQ4Yml0cyBhbmQgRUNDIG9mIGF0IGxlYXN0IDI1NmJpdHMiLCAiaW5jcmVhc2UgcHJpb3JpdHkgb2YgRUNESEUtUlNBLUFFUzI1Ni1HQ00tU0hBMzg0IG92ZXIgREhFLVJTQS1BRVMxMjgtU0hBIiwgImluY3JlYXNlIHByaW9yaXR5IG9mIEFFUzI1Ni1HQ00tU0hBMzg0IG92ZXIgQUVTMTI4LVNIQSIsICJmaXggY2lwaGVyc3VpdGUgb3JkZXJpbmcsIHVzZSByZWNvbW1lbmRlZCBpbnRlcm1lZGlhdGUgY2lwaGVyc3VpdGUiXX19"
        }
    ]
```
I think it would be cleaner to return them as json instead, and we can use the worker name to unmarshal them into the right struct.

**Comment by 0xDiBa** [comment]
Oops. You got me, I had not tested it on a live setup.
I changed []byte to json.RawMessage and everything seems to be ok now.

**Comment by jvehent** [comment]
:v: 
```
$ tlsobs -observatory http://apistage201512031437-2sjrnmheum.elasticbeanstalk.com accounts.firefox.com

--- Certificate ---
Subject  CN=accounts.firefox.com	
SubjectAlternativeName
- accounts.firefox.com
- api.accounts.firefox.com
- scrypt.accounts.firefox.com
- verifier.accounts.firefox.com
- oauth.accounts.firefox.com
- profile.accounts.firefox.com
Issuer   CN=DigiCert SHA2 Secure Server CA
Validity 2014-04-09T00:00:00Z to 2017-04-12T12:00:00Z
CA       false
SHA1     E4289BE190C8DCAE983446B44CEAE3C32FED7465
SHA256   578CC616243D5B7AE48D83F834D498693B5C7ED5E9C406A9985CAB5FB7C9807B
SigAlg   SHA256WithRSA
........................

--- Ciphers Evaluation ---
prio cipher                      protocols             pfs                curves
1    ECDHE-RSA-AES128-GCM-SHA256 TLSv1.2               ECDH,P-256,256bits prime256v1
2    ECDHE-RSA-AES128-SHA256     TLSv1.2               ECDH,P-256,256bits prime256v1
3    ECDHE-RSA-AES128-SHA        TLSv1,TLSv1.1,TLSv1.2 ECDH,P-256,256bits prime256v1
4    DHE-RSA-AES128-SHA          TLSv1,TLSv1.1,TLSv1.2 DH,1024bits        
5    ECDHE-RSA-AES256-GCM-SHA384 TLSv1.2               ECDH,P-256,256bits prime256v1
6    ECDHE-RSA-AES256-SHA384     TLSv1.2               ECDH,P-256,256bits prime256v1
7    ECDHE-RSA-AES256-SHA        TLSv1,TLSv1.1,TLSv1.2 ECDH,P-256,256bits prime256v1
8    AES128-GCM-SHA256           TLSv1.2               None               
9    AES128-SHA256               TLSv1.2               None               
10   AES128-SHA                  TLSv1,TLSv1.1,TLSv1.2 None               
11   AES256-GCM-SHA384           TLSv1.2               None               
12   AES256-SHA256               TLSv1.2               None               
13   AES256-SHA                  TLSv1,TLSv1.1,TLSv1.2 None               
14   DES-CBC3-SHA                TLSv1,TLSv1.1,TLSv1.2 None               
OCSP Stapling        false
Server Side Ordering true
Curves Fallback      false

--- Analyzers ---
	Mozilla evaluation level: intermediate with bad ordering
```

**Pull request merged**